src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/service/SshGitLocationResolverService.java

```diff
@@ -0,0 +1,88 @@
+import de.tum.in.www1.artemis.service.connectors.localvc.LocalVCServletService;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+import de.tum.in.www1.artemis.web.rest.repository.RepositoryActionType;
+
+@Profile(PROFILE_LOCALVC)
+@Service
+public class SshGitLocationResolverService implements GitLocationResolver {
+
+    private static final Logger log = LoggerFactory.getLogger(SshGitLocationResolverService.class);
+
+    @Value("${artemis.version-control.url}")
+    private URL localVCBaseUrl;
+
+    private final LocalVCServletService localVCServletService;
+
+    private final ProgrammingExerciseRepository programmingExerciseRepository;
+
+    public SshGitLocationResolverService(LocalVCServletService localVCServletService, ProgrammingExerciseRepository programmingExerciseRepository) {
+        this.localVCServletService = localVCServletService;
+        this.programmingExerciseRepository = programmingExerciseRepository;
+    }
+
+    @Override
+    public Path resolveRootDirectory(String command, String[] args, ServerSession session, FileSystem fs) throws IOException {
+        // Note: we need to double check read / write access based
+
+        String repositoryPath = args[1];
+        // We need to remove the '/git/' in the beginning
+        if (repositoryPath.startsWith("/git/")) {
+            repositoryPath = repositoryPath.substring(5);
+        }
+
+        String gitCommand = args[0];
+        User user = session.getAttribute(USER_KEY);
+        LocalVCRepositoryUri localVCRepositoryUri = new LocalVCRepositoryUri(Paths.get(repositoryPath), localVCBaseUrl);
+        String projectKey = localVCRepositoryUri.getProjectKey();
+        String repositoryTypeOrUserName = localVCRepositoryUri.getRepositoryTypeOrUserName();
+        ProgrammingExercise exercise;
+
+        try {
+            exercise = programmingExerciseRepository.findOneByProjectKeyOrThrow(projectKey, true);
+        }
+        catch (EntityNotFoundException e) {
+            throw new LocalVCInternalException("Could not find single programming exercise with project key " + projectKey, e);
+        }
+
+        // git-upload-pack means fetch (read operation), git-receive-pack means push (write operation)
+        var repositoryAction = gitCommand.equals("git-upload-pack") ? RepositoryActionType.READ : gitCommand.equals("git-receive-pack") ? RepositoryActionType.WRITE : null;
+        try {
+            localVCServletService.authorizeUser(repositoryTypeOrUserName, user, exercise, repositoryAction, localVCRepositoryUri.isPracticeRepository());
+        }
+        catch (LocalVCForbiddenException e) {
+            log.error("User {} does not have access to the repository {}", user.getLogin(), repositoryPath);
+            throw new AccessDeniedException("User does not have access to this repository", e);
+        }
+
+        try (Repository repo = localVCServletService.resolveRepository(repositoryPath)) {
```

Comment:
```suggestion
        // we cannot trust unvalidated user input
        String localRepositoryPath = localVCRepositoryUri.getRelativeRepositoryPath().toString();
        try (Repository repo = localVCServletService.resolveRepository(localRepositoryPath)) {
```

Reply:
That's actually a good point. I added your suggestion